### ISSUE TITLE
Make wavefd API Object Oriented

### ISSUE THREAD

**Issue by leouieda** [fatiando/fatiando#70]
I have very little taste for OOP and generally prefer functions over classes whenever possible. This is how most of the forward modeling modules in Fatiando work (like `fatiando.gravmag.prism`). So this is how I implemented the initial `fatiando.seismic.wavefd`. It has functions that take the velocity model and simulation parameters and produce iterators for the simulation. It works well enough so far.

The problems with this approach:
- requires a lot of function arguments
- verbose to setup
- hard to automate animation in matplotlib in a general way
- **very** difficult to persist a simulation (using pickle) and resume it later (the user needs to know what to pickle and things can get messy)

This last point is what bugs me the most. Long running simulations would benefit from saving checkpoints that can be used to resume at later times or if things crash.

One way to easily resolve most of these issues is to move to an Object Oriented approach. That way, each type of simulation (acoustic, SH, P-SV, 3D-acoustic) would have its own class. The class would be in charge of setting up the simulation parameters (time step, sources, recording stations, saving snapshots to file, etc.) and velocity model (maybe even include conversion to Lamé parameters?). Each class would have the following methods:
- `run`: run the simulation for a given amount of time (starting from the time step)
- `step`: take a single time step
- `animate`: make a plot animating the simulation using matplotlib.animate

There would be no real need for complex class hierarchy because most methods would have to be reimplemented anyway since they are problem specific.

**Update**: I've made a [prototype of these classes](http://nbviewer.ipython.org/github/fatiando/prototypes/blob/master/wavefd-simulation-class.ipynb) on an IPython notebook.  This is the list of things that I'll try to do soon:
- [x] Store the source as pickled string in HDF5
- [x] Use Abstract Base Classes for WaveFD2D
- [x] Fix problem with plots on non-square domains
- [x] Better use of HDF5 attributes to store metadata (should store the Fatiando version string as well)
- [ ] Set sampling rate and don't store all time steps to disk

After these things are done in the prototype, I'll put them in a branch and make a PR.

**Comment by leouieda** [comment]
After discussion with @eusoubrasileiro, we're going to copy the work done in #137 onto a new branch to get rid of the clutter on that PR. We'll take the lessons learned from there and try to make a better design before writing any more code.

A current list of things to change:

* Make `wavefd` a package to reduce file sizes
* Import the useful classes in `__init__.py`
* Standardize the metadata storage in the HDF5 files
* Move wavelet classes to a new module so that they can be reused elsewhere (like `seismic.conv`)
* Move some of the plotting code to `vis`
* Try to get rid of Cython compiled code (maybe with clever numpy indexing or numba or both)?
* Tests!
* Accept a `SquareMesh` object instead of numpy arrays  to make it  more like other Fatiando code?

**Comment by leouieda** [comment]
#259 discusses an item on the wishlist for wavefd: a method for extracting seismograms that interpolates the wavefield on to specific points (instead of getting the closes node).

**Comment by eusoubrasileiro** [comment]
wow after a lot of time I am getting back to this! I am using the old psv code again....

I suggest we sit down some day to try to this together. The design part isn't a easy task...
I will write down here a little suggestion for future use: "separate the embedding jupyter part from the main code".

**Comment by leouieda** [comment]
Yep, I have to agree on that. I've been thinking about this lately and I'll have to get this done soon for my classes. So it finally became a priority again.
